core/src/main/java/org/elasticsearch/index/seqno/GlobalCheckpointTracker.java

```diff
@@ -249,6 +251,79 @@ public synchronized void updateAllocationIdsFromMaster(
+         * contained in the primary context but not contained in the cluster state applied on the target.
+         *
+         * Suppose there is such a shard and that it is an in-sync shard. However, marking a shard as in-sync requires an operation permit
+         * on the primary shard. Such a permit can not be obtained after the relocation handoff has started as the relocation handoff blocks
+         * all operations. Therefore, there can not be such a shard that is marked in-sync.
+         *
+         * Now consider the case of an initializing shard that is contained in the primary context but not contained in the cluster state
+         * applied on the target.
+         *
+         * If version(context) < version(target) it means that the shard has been removed by a later cluster state update that is already
+         * applied on the target and we only need to ensure that we do not add it to the tracking map on the target. The call to
+         * GlobalCheckpointTracker#updateLocalCheckpoint(String, long) is a no-op for such shards and this is safe.
+         *
+         * If version(target) < version(context) it means that the shard has started initializing by a later cluster state update has not
+         * yet arrived on the target. However, there is a delay on recoveries before we ensure that version(source) <= version(target).
+         * Therefore, such a shard can never initialize from the relocation source and will have to await the handoff completing. As such,
+         * these shards are not problematic.
+         *
+         * Now we consider shards that are contained in the cluster state applied on the target but not contained in the primary context.
+         *
+         * If version(context) < version(target) it means that the target has learned of an initializing shard that the source is not aware
+         * of. As explained above, this initialization can only succeed after the relocation is complete, and only with the target as the
+         * source of the recovery.
+         *
+         * Otherwise, if version(target) < version(context) it only means that the global checkpoint on the target will be held back until a
+         * later cluster state update arrives because the target will not learn of the removal until later.
+         *
+         * In both cases, no calls to update the local checkpoint for such shards will be made. This case is safe too.
+         */
+        for (final ObjectLongCursor<String> cursor : seqNoPrimaryContext.inSyncLocalCheckpoints()) {
+            updateLocalCheckpoint(cursor.key, cursor.value);
+            assert cursor.value >= globalCheckpoint
+                    : "local checkpoint [" + cursor.value + "] violates being at least the global checkpoint [" + globalCheckpoint + "]";
+            try {
+                markAllocationIdAsInSync(cursor.key, cursor.value);
```

Comment:
> I would prefer the sequence number related state of the primary context look like the state of the global checkpoint tracker

Ok. I would still prefer not using markAllocationIdAsInSync as is and all the complexity it brings.